File: 11/4.py
# AP 11.4 - Knight's Tour
import sys

sys.setrecursionlimit(10**7)


def main():
    input = sys.stdin.readline
    N, M = map(int, input().split())
    T = int(input())
    starts = [tuple(map(int, input().split())) for _ in range(T)]
    moves = [(2, 1), (1, 2), (-1, 2), (-2, 1), (-2, -1), (-1, -2), (1, -2), (2, -1)]

    # Count circuits from (0,0)
    sx0, sy0 = 0, 0
    visited = [[False] * M for _ in range(N)]
    visited[sx0][sy0] = True
    circuits = 0

    def dfs_circ(x, y, depth):
        nonlocal circuits
        if depth == N * M:
            for dx, dy in moves:
                if x + dx == sx0 and y + dy == sy0:
                    circuits += 1
                    break
            return
        for dx, dy in moves:
            nx, ny = x + dx, y + dy
            if 0 <= nx < N and 0 <= ny < M and not visited[nx][ny]:
                visited[nx][ny] = True
                dfs_circ(nx, ny, depth + 1)
                visited[nx][ny] = False

    dfs_circ(sx0, sy0, 1)
    print(circuits)

    def count_paths(sx, sy):
        visited2 = [[False] * M for _ in range(N)]
        visited2[sx][sy] = True
        cnt = 0

        def dfs(x, y, depth):
            nonlocal cnt
            if depth == N * M:
                cnt += 1
                return
            for dx, dy in moves:
                nx, ny = x + dx, y + dy
                if 0 <= nx < N and 0 <= ny < M and not visited2[nx][ny]:
                    visited2[nx][ny] = True
                    dfs(nx, ny, depth + 1)
                    visited2[nx][ny] = False

        dfs(sx, sy, 1)
        return cnt

    for sx, sy in starts:
        print(count_paths(sx, sy))


if __name__ == "__main__":
    main()

File: 11/4.cpp
// AP 11.4 - Knight's Tour
#include <iostream>
#include <vector>
#include <functional>
using namespace std;

int N, M;
int circuits = 0;
vector<pair<int, int>> moves = {{2, 1}, {1, 2}, {-1, 2}, {-2, 1}, {-2, -1}, {-1, -2}, {1, -2}, {2, -1}};

void dfs_circuit(int x, int y, int depth, vector<vector<bool>> &visited)
{
    if (depth == N * M)
    {
        for (auto &move : moves)
        {
            int nx = x + move.first;
            int ny = y + move.second;
            if (nx == 0 && ny == 0)
            {
                circuits++;
                break;
            }
        }
        return;
    }

    for (auto &move : moves)
    {
        int nx = x + move.first;
        int ny = y + move.second;
        if (nx >= 0 && nx < N && ny >= 0 && ny < M && !visited[nx][ny])
        {
            visited[nx][ny] = true;
            dfs_circuit(nx, ny, depth + 1, visited);
            visited[nx][ny] = false;
        }
    }
}

int count_paths(int sx, int sy)
{
    vector<vector<bool>> visited(N, vector<bool>(M, false));
    visited[sx][sy] = true;
    int cnt = 0;

    function<void(int, int, int)> dfs = [&](int x, int y, int depth)
    {
        if (depth == N * M)
        {
            cnt++;
            return;
        }

        for (auto &move : moves)
        {
            int nx = x + move.first;
            int ny = y + move.second;
            if (nx >= 0 && nx < N && ny >= 0 && ny < M && !visited[nx][ny])
            {
                visited[nx][ny] = true;
                dfs(nx, ny, depth + 1);
                visited[nx][ny] = false;
            }
        }
    };

    dfs(sx, sy, 1);
    return cnt;
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> N >> M;
    int T;
    cin >> T;

    vector<pair<int, int>> starts(T);
    for (int i = 0; i < T; i++)
    {
        cin >> starts[i].first >> starts[i].second;
    }

    // Count circuits from (0,0)
    vector<vector<bool>> visited(N, vector<bool>(M, false));
    visited[0][0] = true;
    circuits = 0;
    dfs_circuit(0, 0, 1, visited);
    cout << circuits << "\n";

    // Count paths for each starting point
    for (auto &start : starts)
    {
        cout << count_paths(start.first, start.second) << "\n";
    }

    return 0;
}